pkg/resource/api.go

```diff
@@ -47,90 +47,118 @@ func NewAPIPatchingApplicator(c client.Client) *APIPatchingApplicator {
 	return &APIPatchingApplicator{client: c}
 }
 
+// WithLogger sets the logger on the APIPatchingApplicator.
+func (a *APIPatchingApplicator) WithLogger(l logging.Logger) *APIPatchingApplicator {
+	a.optionalLog = l
+	return a
+}
+
 // Apply changes to the supplied object. The object will be created if it does
 // not exist, or patched if it does. If the object does exist, it will only be
 // patched if the passed object has the same or an empty resource version.
-func (a *APIPatchingApplicator) Apply(ctx context.Context, o client.Object, ao ...ApplyOption) error {
-	m, ok := o.(metav1.Object)
-	if !ok {
-		return errors.New("cannot access object metadata")
-	}
-
-	if m.GetName() == "" && m.GetGenerateName() != "" {
-		return errors.Wrap(a.client.Create(ctx, o), "cannot create object")
+func (a *APIPatchingApplicator) Apply(ctx context.Context, obj client.Object, ao ...ApplyOption) error { //nolint:gocyclo // the logic here is crucial and deserves to stay in one method
+	if obj.GetName() == "" && obj.GetGenerateName() != "" {
+		return errors.Wrap(a.client.Create(ctx, obj), "cannot create object")
 	}
 
-	desired := o.DeepCopyObject()
-
-	err := a.client.Get(ctx, types.NamespacedName{Name: m.GetName(), Namespace: m.GetNamespace()}, o)
+	current := obj.DeepCopyObject().(client.Object)
+	err := a.client.Get(ctx, types.NamespacedName{Name: obj.GetName(), Namespace: obj.GetNamespace()}, current)
 	if kerrors.IsNotFound(err) {
 		// TODO(negz): Apply ApplyOptions here too?
-		return errors.Wrap(a.client.Create(ctx, o), "cannot create object")
+		return errors.Wrap(a.client.Create(ctx, obj), "cannot create object")
 	}
 	if err != nil {
 		return errors.Wrap(err, "cannot get object")
 	}
 
+	// Note: this check would ideally not be necessary if the Apply signature
+	// had a current object that we could us for the diff. But we have no
+	// current and for consistency of the patch it matters that the object we
+	// get above is the one that was originally used.
+	if obj.GetResourceVersion() != "" && obj.GetResourceVersion() != current.GetResourceVersion() {
+		gvr, err := groupResource(a.client, obj)
+		if err != nil {
+			return err
+		}
+		return kerrors.NewConflict(gvr, current.GetName(), errors.New("resource version does not match"))
+	}
+
 	for _, fn := range ao {
-		if err := fn(ctx, o, desired); err != nil {
+		if err := fn(ctx, current, obj); err != nil {
 			return err
 		}
 	}
 
+	if err := LogDiff(a.optionalLog, current, obj); err != nil {
```

Comment:
given that diff compute is expensive, could we avoid doing it twice, in case when `optionalLog` is set? It is computed by `client.MergeFromWithOptions(current, client.MergeFromWithOptimisticLock{}))` anyway, so perhaps we can create a patch wrapper around it to cache the computed value?

Reply:
good idea. done